mirrord-layer/src/file.rs

```diff
@@ -0,0 +1,170 @@
+// https://doc.rust-lang.org/std/panic/fn.catch_unwind.html
+
+// TODO(alex) [low] 2022-05-03: Safe wrappers around these functions.
+
+// https://users.rust-lang.org/t/problem-overriding-libc-functions-with-ld-preload/41516/4
+
+/// NOTE(alex): Regex that ignores system files + files in the current working directory.
+static IGNORE_FILES: SyncLazy<RegexSet> = SyncLazy::new(|| {
+    // NOTE(alex): To handle the problem of injecting `open` and friends into project runners (like
+    // in a call to `node app.js`, or `cargo run app`), we're ignoring files from the current
+    // working directory (as suggested by aviram).
+    let mut cwd_buffer = [0; libc::PATH_MAX as usize];
+    let cwd = unsafe { libc::getcwd(cwd_buffer.as_mut_ptr(), cwd_buffer.len()) };
+
+    let cwd_str = unsafe { CStr::from_ptr(cwd) }
+        .to_str()
+        .expect("Failed converting cwd from c_char!");
+
+    let set = RegexSet::new(&[
+        r".*\.so",
+        r".*\.d",
+        r"^/proc/.*",
+        r"^/sys/.*",
+        r"^/lib/.*",
+        r"^/etc/.*",
+        r"^/usr/.*",
+        r"^/dev/.*",
+        // TODO(alex) [low] 2022-05-13: Investigate why it's trying to load this file all of a
+        // sudden. We have to ignore this here as `node` will look for it outside the cwd (it
+        // searches up the directory, until it finds).
+        r".*/package.json",
+        cwd_str,
+    ])
+    .unwrap();
+
+    set
+});
+
+/// Blocking wrapper around `libc::open` call.
+///
+/// It's bypassed when trying to load system files, and files from the current working directory
+/// (which is different anyways when running in `-agent` context).
+///
+/// When called for a valid file, it'll block, send a `ClientMessage::OpenFileRequest` to be handled
+/// by `mirrord-agent` (`handle_peer_message` function), and wait until it receives a
+/// `DaemonMessage::FileOpenResponse`.
+pub fn open(raw_path: *const c_char, oflag: c_int) -> RawFd {
+    let path: PathBuf = unsafe { CStr::from_ptr(raw_path) }
+        .to_str()
+        .expect("Failed converting path from c_char!")
+        .into();
+
+    if IGNORE_FILES.is_match(path.to_str().unwrap()) {
+        let fd = unsafe { libc::open(raw_path, oflag) };
+        fd
+    } else {
+        debug!("open -> trying to open valid file {path:?}");
+        let sender = unsafe { HOOK_SENDER.as_ref().unwrap() };
+        let (file_channel_tx, file_channel_rx) = oneshot::channel::<FileOpenResponse>();
+
+        let requesting_file = OpenFileHook {
+            path,
+            file_channel_tx,
+        };
+
+        match sender.blocking_send(HookMessage::OpenFileHook(requesting_file)) {
+            Ok(_) => {}
+            Err(fail) => {
+                error!("open: failed to send open message: {fail:?}!");
+                return libc::EFAULT;
+            }
+        };
+
+        debug!("Blocking on file open operation!");
+
+        let file_open = RUNTIME.block_on(async {
```

Comment:
I'm pretty sure we don't want to use tokio runtime in hooks? I didn't actually think that's a possibility, so maybe we do and it'll make rest of the code easier..?
anyway you can do https://docs.rs/tokio/latest/tokio/sync/oneshot/struct.Receiver.html#method.blocking_recv

Reply:
Think both are roughly equivalent, but I'm changing it to use `tokio::oneshot` instead of `futures::oneshot` and apply your suggestion of using `blocking_recv`.